### ISSUE TITLE
[Feature Request] Customisable connection icons

### ISSUE THREAD

**Issue by antoineraulin** [xpipe-io/xpipe#335]
Hi,

I'd like to propose a new feature that could enhance user experience when handling multiple connections (SSH or other) within your app. Right now, it can get confusing with many machines or VMs to tell apart different connections.

What if we added an option to customise connection icons, like MobaXterm does? Users could set unique icons for each connection based on the service hosted by the machine. This would make identifying active connections easier and quicker.

If technically possible with Microsoft's Console API, applying these custom icons to Windows Terminal tab bar when using XPipe would be super beneficial. While you can define personalised icons while creating profiles in Windows Terminal, I'm not sure if this is achievable through the API.

I understand this might cater to a specific niche of users, but I think it could still be an interesting and useful addition to your project.

Best,

**Comment by crschnick** [comment]
So you're looking for having a set of icons to choose from?

Also do you have a link for applying icons to Windows Terminal like you mentioned?

**Comment by antoineraulin** [comment]
Not necessarily a set of icons. I'm just asking for the ability to choose a PNG or SVG file from my PC that would replace the SSH icon next to the connection name.

However, a set of icons could be a good idea. The downside is that unless the set is very generic (icons for PC, server, proxy, firewall, etc.), logos of popular services can quickly become outdated if not regularly updated. Plus, it’s often not exhaustive, especially when dealing with niche open-source services.

As for the question regarding Windows Terminal, the functionality for setting an icon on profiles is explained here: https://learn.microsoft.com/en-us/windows/terminal/customize-settings/profile-appearance#icons. 

It looks like this in the terminal :
![{B9FCD273-83F8-4B81-84D5-2074227439BE}](https://github.com/user-attachments/assets/e563e091-684d-425c-8296-8658ca76381f)

**Comment by crschnick** [comment]
With Windows terminal that is going to be difficult as custom icons are not supported from the CLI and require the creation of a custom profile.

But worked I can also think of is recognizing common types of systems and containers. E.g. if xpipe would recognize that you connected to a traefix container, it could show that icon and name for traefik instead of the normal os name/logo.

**Comment by crschnick** [comment]
I could a chooser for the following icons: https://kordamp.org/ikonli/cheat-sheet-material2.html

They are already included in xpipe.

**Comment by antoineraulin** [comment]
Detecting the service running on a container could indeed be an approach. However, I don't know how it would work if the target machine was running several services or if it wasn't a container. I'll take my own case as an example: I use almost no Docker containers (partly because they still don't handle IPV6 very well) and therefore have as many Proxmox VMs/CTs as services. 

Especially as it means maintaining a hard-coded list in XPipe of services -> script to detect it.

I imagine that integrating custom icons isn't necessarily easy or involves refactoring some of the code.

**Comment by crschnick** [comment]
Yeah I see that detecting the running service would be too much work and still wouldn't cover all cases.

Integrating custom icons wouldn't be too difficult. I think the main challenge would be that the available icons are too many: https://kordamp.org/ikonli/cheat-sheet-material2.html

E.g. it could look something like this 
![image](https://github.com/user-attachments/assets/8ded3da7-deab-4137-8b59-06ffeabaf51f)

**Comment by antoineraulin** [comment]
I agree, materials icons are extremely varied and numerous, not to mention the fact that they don't have any particular IT focus, they're mainly UI interaction icons for applications.

Allowing only the user to provide their own icons as png files would solve the problem: the user is free to choose and there's a lot less maintenance on your side.

**Comment by crschnick** [comment]
The problem is that with custom images the resolution does not match and the image quality will not look as good. Also, ideally the images should work in both light and dark mode.

**Comment by antoineraulin** [comment]
But wouldn't that be the user's problem? If they want to put up an ugly image, so be it. There's no obligation to use the feature. 

I completely understand the desire for visual consistency.

Alternatively, there are SVG icon libraries dedicated to this kind of subject: https://techicons.dev/ but there are still a lot of icons missing (I'm thinking of Caddy Web Server, Paperless, Bind NS, etc.).

**Comment by antoineraulin** [comment]
For the image quality problem, we could imagine allowing only SVGs

**Comment by crschnick** [comment]
Yes, but rendering svgs is expensive. So that is not really feasible to implement

**Comment by crschnick** [comment]
So the only thing that is really possible is adding the ability to use custom images which won't look that good as they probably have scaling artifacts. And it's not easily possible to use them in terminals, so they would only show up in the connection hub.

**Comment by antoineraulin** [comment]
If it's not displayed in the terminal, that's fine, it's less annoying, it's quite rare to have 45 different tabs open.On the other hand, when it comes to the list of connections, it's difficult for me to segment any further and I always have several dozen lines. For the scaling issue, could we imagine the file picker indicating the accepted dimensions to avoid it? I admit that this would increase the friction of using the functionality.Le 29 août 2024 à 15:12, Christopher Schnick ***@***.***> a écrit :﻿
So the only thing that is really possible is adding the ability to use custom images which won't look that good as they probably have scaling artifacts. And it's not easily possible to use them in terminals, so they would only show up in the connection hub.

**Comment by crschnick** [comment]
I also thought about how to handle synced connections, because then the images would also need to be committed to the git repository. Overall this is all very complex.

I think if you say that it is difficult for you to segment any further, I can think of other measures to allow for an easier overview and organization if you have many connections. And better ways to locate a specific connection with little effort.

**Comment by antoineraulin** [comment]
I hadn't thought about sync. 

If you ever find a way to find your way around better, great. Otherwise it's no big deal

**Comment by crschnick** [comment]
I found this: https://selfh.st/icons/

**Comment by crschnick** [comment]
Well would these icons cover most of your services that you are running?

**Comment by antoineraulin** [comment]
Pretty much yes, selfh.st is a good repository of what's going on in the selfhost world and they add new services to their database fairly regularly. I only knew them for the service repository but not for the icons.

**Comment by crschnick** [comment]
Like this?

![image](https://github.com/user-attachments/assets/9c9ea356-ee42-412a-a160-ff3ea855cb01)

**Comment by antoineraulin** [comment]
If this feature is a hit with users, I wouldn't be surprised if they were involved in the development of automatic detection.  
For detection at container level, it seems pretty straightforward to me, detection at image name level is enough, but for more exotic installations, I don't know how it would work. Sometimes, installation guides indicate precise locations for certain files, but nothing obliges the user to follow them. But I imagine you can be much more creative than that when it comes to detection (searching for systemd services, or well-known processes). To tell you the truth, I don't know how detection usually works in XPipe.

In any case - if I understand how it works - I'll try to implement it for the services I use.

regarding the picker, I suppose a search bar could improve the UX. The selfh.st icon base is pretty big, isn't it?

**Comment by crschnick** [comment]
This is now implemented in the PTB.

You can find the implementation for the detection logic at https://github.com/xpipe-io/xpipe/blob/737d5ce68c116863d26acc9376ef27fd9f375917/app/src/main/java/io/xpipe/app/resources/SystemIcons.java#L36

It is currently only detecting filebrowser to test.

**Comment by antoineraulin** [comment]
Thank you for implementing the initial version ! I’ve tested it out with the latest Public Test Beta, but I’m having some trouble getting the automatic detection for "File Browser" to work.

To test, I created two Docker containers: one named "filebrowser" and the other by simply launching the `filebrowser/filebrowser` Docker image. Unfortunately, neither seems to trigger the detection. I might be doing something wrong — is there any specific setup or configuration I should follow to make it work? Are there logs or checks I can perform to verify if the detection is being attempted?

Also, I imagine the icon picker functionality isn’t implemented yet, but I couldn’t find an option to manually select an icon either. If there’s anything I can try or if I’ve missed something in the settings, feel free to let me know.

Again, I really appreciate the time and effort you’re putting into this.

![image](https://github.com/user-attachments/assets/ce1f7a16-2f8b-45cf-aaf9-8dee69cd21bf)

**Comment by crschnick** [comment]
You can change icons manually by clicking on the current icon for a specific connection, maybe I need to make that more clear somehow.

About the detection of container images, what happens if you click on refresh for the docker default context? That works for me.

About the local machine OS name, that looks wrong. Was it always like that?

**Comment by antoineraulin** [comment]
I started XPipe this morning, and the icons for File Browser appeared automatically, so I didn’t get a chance to test the manual refresh method you suggested, but it seems to be working fine now.

As for the icon picker, it works perfectly! I had initially looked for the option in the menu, similar to changing the connection color, but once I clicked on the icon itself, everything worked smoothly. Perhaps making that a bit clearer in the UI would help other users, but the feature itself is great.
![image](https://github.com/user-attachments/assets/2caffc92-5c61-4ea4-ba03-e915c1e3a2a3)

Regarding the OS name display issue, I don’t typically see it because I usually have folders on the right side and often hide the host machine. However, I believe it’s been that way since I first installed the app on this machine. For reference, the language and region settings are set to French/France.

Lastly, I’ve noticed that on this PC, custom icons and the default ones look quite pixelated. This might be related to my 4K display with scaling set to 200% in Windows. I don’t use this machine as often (I’m currently on vacation), so I’m not sure if this issue was present before, but I thought it might be worth mentioning :
![image](https://github.com/user-attachments/assets/cab80cde-d6b2-4293-951e-296e6f8e34fc)
![image](https://github.com/user-attachments/assets/d05c42d1-a59f-4def-9d13-cfd1f3b624d6)

**Comment by crschnick** [comment]
Good to know that it works, I can make the picker more intuitive to find.

The image resolution is due to your scaling, that is a limitation in the approach. I don't think I can fix that.

That os name issue might actually look small, but that is worrying for me because that shouldn't happen. This might cause other issues down the line if there are encoding issues.

Can you try to run the following commands in cmd:

```
chcp
wmic os get Caption /format:value
wmic os get Caption /format:value > out1.txt
chcp 65001
wmic os get Caption /format:value
wmic os get Caption /format:value > out2.txt
```
and share the output and both output files?

**Comment by antoineraulin** [comment]
Thanks for the clarification! No worries about the image resolution, it’s not a big issue for me.

I ran the commands you provided, and I’ve attached both output files. There don’t seem to be any encoding artifacts in them.

[out2.txt](https://github.com/user-attachments/files/17010905/out2.txt)
[out1.txt](https://github.com/user-attachments/files/17010907/out1.txt)

**Comment by crschnick** [comment]
Can you also share the console output when you run these commands? Preferable in a file as GitHub might change some characters

**Comment by crschnick** [comment]
Edit: I learned something new today. Seems like wmic changes the output encoding based on whether it writes to a file or not.

So to get the original output, it needs to be something like this:

```
wmic os get Caption /format:value | find /v "" > out1.txt
chcp 65001
wmic os get Caption /format:value | find /v "" > out2.txt
```

**Comment by antoineraulin** [comment]
The file with the console output :

[PS CUsersantoine chcp.txt](https://github.com/user-attachments/files/17010983/PS.CUsersantoine.chcp.txt)

**Comment by antoineraulin** [comment]
The new out1 et 2 are empty due to those errors :
```console
PS C:\Users\antoine> wmic os get Caption /format:value | find /v "" > out1.txt
FIND : format incorrect de paramètre
PS C:\Users\antoine> chcp 65001
Page de codes active : 65001
PS C:\Users\antoine> wmic os get Caption /format:value | find /v "" > out2.txt
FIND : format incorrect de paramètre
PS C:\Users\antoine>
```

[out1.txt](https://github.com/user-attachments/files/17010988/out1.txt)
[out2.txt](https://github.com/user-attachments/files/17010989/out2.txt)

**Comment by antoineraulin** [comment]
> And you need to run this in cmd.exe not powershell as xpipe does as well

Ok will do

**Comment by antoineraulin** [comment]
> wmic os get Caption /format:value > out2.txt
With cmd :
[CUsersantoinechcp.txt](https://github.com/user-attachments/files/17011019/CUsersantoinechcp.txt)

[out1.txt](https://github.com/user-attachments/files/17011014/out1.txt)
[out2.txt](https://github.com/user-attachments/files/17011015/out2.txt)

**Comment by antoineraulin** [comment]
> wmic os get Caption /format:value | find /v "" > out1.txt
> chcp 65001
> wmic os get Caption /format:value | find /v "" > out2.txt
Also with cmd :
[out1.txt](https://github.com/user-attachments/files/17011029/out1.txt)
[out2.txt](https://github.com/user-attachments/files/17011031/out2.txt)

**Comment by crschnick** [comment]
Hmm, so am I correct that if you run that in powershell it displays the accent on the `E` but if you run the same thing in cmd, it does not?

**Comment by antoineraulin** [comment]
Indeed, but XPipe should see without the accent then ?

**Comment by crschnick** [comment]
Yes, it should. I think in your test with cmd you didn't run that in a new session, so it took the modified chcp in both cases,

Can you run
```
chcp 850
wmic os get Caption /format:value | find /v "" > out1.txt
```
again and share the file?

**Comment by crschnick** [comment]
I think this a Windows bug. Looking at the output in a hex editor for example the powershell output, there are some weird chars in there: 

![image](https://github.com/user-attachments/assets/c80f1d32-2d01-446b-a37a-f09e161df28a)

The accent is not the only issue and I don't know where that char between Windows and 11 comes from.

**Comment by crschnick** [comment]
Now for the reason why I think this might be problematic is that if the issue also occurs in other areas.

If you open the xpipe file browser for the local machine and create a new file with special characters such as accents, does it create a file the correct name?

**Comment by antoineraulin** [comment]
> Yes, it should. I think in your test with cmd you didn't run that in a new session, so it took the modified chcp in both cases,
> 
> 
> 
> Can you run
> 
> ```
> 
> chcp 850
> 
> wmic os get Caption /format:value | find /v "" > out1.txt
> 
> ```
> 
> again and share the file?

Will do after lunch :)

**Comment by antoineraulin** [comment]
So :
```console
chcp 850

wmic os get Caption /format:value | find /v "" > out1.txt
```
done in a new terminal, gave no encoding issue :
![image](https://github.com/user-attachments/assets/eaf0465c-e1d2-4b2e-abe3-f400210d79a1)
[out1.txt](https://github.com/user-attachments/files/17011757/out1.txt)

**Comment by crschnick** [comment]
can you create files in the xpipe file browser with special characters such as accents?

**Comment by antoineraulin** [comment]
No issues regarding the creation of files with special chars : 
![image](https://github.com/user-attachments/assets/a8391a19-267d-45fc-9c2e-d67ed11a3d03)
I could edit it and save it through XPipe with no problem

**Comment by antoineraulin** [comment]
Yes, that's a file created through XPipe in the file browser feature on the windows host local machine.

I can't blame you, wmic is such a pain to use when dealing with different host languages

**Comment by crschnick** [comment]
Yeah I can't use powershell in that context. Mainly because I can't assume that it works but also because it's very slow.

**Comment by crschnick** [comment]
So if you want, you can add the custom icon detection for the services you run.

I can then publish another PTB for that.

**Comment by crschnick** [comment]
Right now yes, but I can quickly add support for other container types as well

**Comment by antoineraulin** [comment]
I ask because most of my services run on LXC. If possible, a way of setting up detection by binary present on the machine would be perfect to accommodate the different ecosystems.

**Comment by crschnick** [comment]
Yeah it can check whether a binary is present but then the icon would only update when you establish a connection to the system.

**Comment by antoineraulin** [comment]
If it's the same behaviour as when XPipe detects the OS of a connection then there's no problem, it seems perfectly normal to me that an initial connection is required to run scripts.

**Comment by crschnick** [comment]
So I added support to check for specific files to determine the right system icon: https://github.com/xpipe-io/xpipe/blob/6eb70ea7e5b0c55017cdb5efb630c732fc92f30d/app/src/main/java/io/xpipe/app/resources/SystemIcons.java#L38

**Comment by antoineraulin** [comment]
Awesome, many thanks

**Comment by crschnick** [comment]
The latest PTB build should fix the OS name display issues and the texture rendering scaling issues

**Comment by crschnick** [comment]
The custom icons feature and fixes for all the issues are now released in xpipe 12.
